docs/best-practices.md

```diff
@@ -0,0 +1,82 @@
+- The open parameters should contain information that is only relevant for opening a specific event source, and their lifecycle ends at the invokation of `plugin_close()`
+- The open parameters should contain minimal and non-structured information, such as an URI or a resource descriptor string. This is the reason why the framework does not support any schema definition for open parameters, and treats them as a opaque string. Ideally, if more than one parameter is required to open a data source, comma-separated string contatenation is preferrable than structured data formats such as JSON
+
+## Secret Management
+
+An important decision during plugin development is how secrets are managed. It's common to manage connections to external services, authenticate with credentials, or use private tokens. Plugins accept user-defined values only through the init configuration or the open parameters. In general, you shouldn't pass secrets directly inside the init configuration or the open parameters. Instead, there are few guidelines you can follow in order to ensure that your secrets are managed safely in the plugin framework.
+
+### Don't
+
+- Pass credentials or secrets in the plugin:
+    - Init configuration
+    - Open parameters
+- Create custom conventions (e.g. storing the secrets in a non-customizable path such as `~/.mysecrets`)
+- Embed credentials or secrets inside the plugin code
+
+### Do
+
+- Retrieve credentials or secrets through environment variables, or external files, and pass their name in the plugin:
+    - Init configuration
+    - Open parameters, only if they are required to open the event source
+- Follow the secret management guidelines imposed by well-known packages such as [aws-sdk-go](https://github.com/aws/aws-sdk-go#configuring-credentials)
+
+
+## Managing the NextBatch Loop
+
+In source plugins, new events are created and returned in batches with the `plugin_next_batch()` function. This design decision reduces the overhead that may occur by integrating C with other programming languages such as Go. Although the SDKs cover most of the hard work, plugin developers must manage the batching system manually and there are some things to keep in mind in order to ensure optimal performance and correct behavior.
+
+### Don't
+
+- Allocate a new event batch at every invokation of `plugin_next_batch()`, because event generation is a very hot path
+- Block the execution for long times: be mindful that `plugin_next_batch()` is blocking for the framework
+- Return an EOF error and then return new events at the following invokation of `plugin_next_batch()`
+- Return a given number of events without filling their data in the event batch
```

Comment:
```suggestion
- Allocate a new event batch at every invocation of `plugin_next_batch()`, because event generation is a very hot path
- Block the execution for long times: be mindful that `plugin_next_batch()` is blocking for the framework
- Return an EOF error and then return new events at the following invocation of `plugin_next_batch()`
- Return a given number of events without filling their data in the event batch
```